File: app/src/main/java/dictionary/Dictionary.java
package dictionary;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Dictionary {

    public static class DictionaryBuilder {
        private final String outputPath;
        private final List<DictionaryPage> pages;

        public DictionaryBuilder(String outputPath) {
            this.outputPath = outputPath;
            pages = new ArrayList<>();
        }

        public Dictionary build() throws IOException {
            return new Dictionary(this);
        }

        public DictionaryBuilder cover(String title, String createdBy) {
            pages.add(new CoverPage(title, createdBy, outputPath));
            return this;
        }

        public DictionaryBuilder content(String contentPath) throws FileNotFoundException {
            pages.add(new ContentPage(outputPath, contentPath));
            return this;
        }

        public DictionaryBuilder copyright() {
            pages.add(new CopyrightPage(outputPath));
            return this;
        }
    }

    private final List<DictionaryPage> pages;

    private Dictionary(DictionaryBuilder builder) throws IOException {
        System.out.println("Creating Custom Kindle Dictionary...");
        pages = builder.pages;
        if (!pages.isEmpty()) {
            for (var page : pages) {
                System.out.println("Creating: " + page.getOutFile().getName());
                page.build();
            }
            System.out.println(
                    "Dictionary files created: " + pages.get(0).getOutFile().getParentFile().getAbsolutePath());
        } else {
            System.out.println("No Files created.");
        }
    }

}

File: app/src/main/java/dictionary/Dictionary.kt
package dictionary

import java.io.FileNotFoundException
import java.io.IOException

/**
 *  Kindle Custom Dictionary pages
 */
class Dictionary private constructor(builder: DictionaryBuilder) {
    /**
     * Builder to create the pages of a custom kindle dictionary
     * @param outputPath path to output the created dictionary pages
     */
    class DictionaryBuilder(private val outputPath: String) {
        private val _pages = arrayListOf<DictionaryPage>()

        val pages: List<DictionaryPage>
            get() = _pages

        @Throws(IOException::class)
        fun build(): Dictionary {
            return Dictionary(this)
        }

        fun cover(title: String, createdBy: String): DictionaryBuilder {
            _pages.add(CoverPage(title, createdBy, outputPath))
            return this
        }

        @Throws(FileNotFoundException::class)
        fun content(contentPath: String): DictionaryBuilder {
            _pages.add(ContentPage(outputPath, contentPath))
            return this
        }

        fun copyright(): DictionaryBuilder {
            _pages.add(CopyrightPage(outputPath))
            return this
        }
    }

    private val pages = builder.pages

    init {
        println("Creating Custom Kindle Dictionary...")
        if (pages.isNotEmpty()) {
            pages.forEach {
                println("Creating: ${it.outFile.name}")
                it.build()
            }
            println(
                "Dictionary files created: ${pages[0].outFile.parentFile.absolutePath}"
            )
        } else {
            println("No Files created.")
        }
    }
}